### DISCUSSION TITLE
CharacterVirtual Colliding at Center of Mass?

### DISCUSSION THREAD

**Discussion by wirepair** [jrouwe/JoltPhysics#1113] [Q&A] [answered]
Got a real weird one, I have a simple CharacterVirtual which is pretty much copied from the Jolt Samples app:
```cpp
float CharacterHeightStanding = 1.35f;
float CharacterRadiusStanding = 0.3f;

StandingShape = JPH::RotatedTranslatedShapeSettings(JPH::Vec3(0, 0.5f * CharacterHeightStanding + CharacterRadiusStanding, 0), JPH::Quat::sIdentity(), new JPH::CapsuleShape(.5f * CharacterHeightStanding, CharacterRadiusStanding)).Create().Get();

Settings = new JPH::CharacterVirtualSettings();
Settings->mMaxSlopeAngle = JPH::DegreesToRadians(45.0f);
Settings->mMaxSlopeAngle = sMaxSlopeAngle;
Settings->mMaxStrength = sMaxStrength;
Settings->mShape = StandingShape;
Settings->mBackFaceMode = sBackFaceMode;
Settings->mCharacterPadding = sCharacterPadding;
Settings->mPenetrationRecoverySpeed = sPenetrationRecoverySpeed;
Settings->mPredictiveContactDistance = sPredictiveContactDistance;
Settings->mSupportingVolume = Plane(Vec3::sAxisY(), -CharacterRadiusStanding); // Accept contacts that touch the lower sphere of the capsule 

PhysicsCharacter = new JPH::CharacterVirtual(Settings, JPH::RVec3::sZero(), JPH::Quat::sIdentity(), 0, PhysicsSystem);
PhysicsCharacter->SetListener(this);
```

and a box for the character to stand on:
```cpp
        auto Floor = BodyInterface.CreateBody(BodyCreationSettings(new JPH::BoxShape(JPH::Vec3(0.5f * 300, 1.0f, 0.5f * 300), 0.0f), JPH::RVec3(JPH::Vec3(0.0f, -1.0f, 0.0f)), JPH::Quat::sIdentity(), JPH::EMotionType::Static, physics::Layers::NON_MOVING));
        auto BodyID = Floor->GetID();
        BodyInterface.AddBody(BodyID, EActivation::DontActivate);
```

and for some reason it seems to collide with the center of mass instead of the bottom of the capsule?

![image](https://github.com/jrouwe/JoltPhysics/assets/1073742/eaa667a6-c6e4-4bd9-baee-8141d31e7a91)

Collision is working fine as it's triggering the ContactListener implementations for the character, yet it seems to be colliding with the center of mass? I've tried changing the `JPH::RotatedTranslatedShapeSettings` position but it doesn't seem to make any difference.

Debug render code is again, the same from the CharacterVirtual sample app:
```cpp
JPH::RMat44 CenterOfMass = PhysicsCharacter->GetCenterOfMassTransform();
auto CharShape = Character.Character->PhysicsCharacter->GetShape();
CharShape->Draw(DebugRenderer.get(), CenterOfMass, JPH::Vec3::sReplicate(100.0f), Color::sGreen, true, true);
```

Yet in the Jolt Samples app the CharacterVirtual clearly stands on top of the ground normally.

Any thoughts on how further to debug or fix are appreciated!

**Comment by wirepair** [comment]
Oddly, I was able to get it to work if I updated the shape using `OffsetCenterOfMassShape` but since the Jolt Sample app doesn't need to do this, I'm worried I'm just doing something wrong and additional problems will show up later on. 
```
StandingShape = new JPH::OffsetCenterOfMassShape(StandingShape, -StandingShape->GetCenterOfMass());
```

**Comment by jrouwe** [comment]
I'm not really sure what I'm looking at. How are you drawing the box? If I look at the screenshot, it looks like the character is standing perfectly on the floor of your level but it appears that the box (assuming this represents the floor) is being drawn in the wrong spot?

Note that the center of mass of the character is the center of the capsule, so `GetCenterOfMassTransform()` is not supposed to return the bottom of the capsule. Use `GetPosition()` or `GetWorldTransform()` for that. To get the actual contact position you'd call `GetGroundPosition()`.

PS1. In your use of `OffsetCenterOfMassShape` you make the center of mass of the capsule coincide with the bottom of the capsule so then `GetCenterOfMassTransform() == GetWorldTransform()`, if it works by doing this, you're probably confusing center of mass transform with the regular transform somewhere.
PS2. `CharShape->Draw` indeed takes `GetCenterOfMassTransform()` (you can see that in the parameter name: `inCenterOfMassTransform`), so the issue is not there.

**Reply by wirepair** [reply]
Sorry, that initial image was terribly mis-leading as I actually removed physics from unreal engine, here's a better picture of the issue:
![image](https://github.com/jrouwe/JoltPhysics/assets/1073742/25ac3c32-ec00-4f5d-99f9-aa8e5422b7a5)

> Note that the center of mass of the character is the center of the capsule, so GetCenterOfMassTransform() is not supposed to return the bottom of the capsule.

That's precisely why I'm surprised to see my character considering a collision with the floor box at the center of the capsule instead of the bottom of it.


I expect it to look like it does in the JoltSamples where the bottom of the capsule is what is touching the top of the box:
![image](https://github.com/jrouwe/JoltPhysics/assets/1073742/d19c3bdb-8aeb-4723-abe3-da90c397e504)

As for how I am printing the box, I'm simply using the DrawBodies call on the PhysicsSystem, do I need to offset that with something?
```cpp
      JPH::BodyManager::DrawSettings BodyDrawSettings;
      BodyDrawSettings.mDrawShapeWireframe = true;
      System->DrawBodies(BodyDrawSettings, DebugRenderer.get(), nullptr);
```

Which feeds into a custom debug render class built for UE5:
```
 virtual void DrawLine(debug::DebugVec3& From, debug::DebugVec3& To, uint32_t InColor) override
    {
        FVector DebugFrom{ From.X, From.Z, From.Y };
        FVector DebugTo{ To.X, To.Z, To.Y };
        // ... color stuff ...
        UKismetSystemLibrary::DrawDebugLine(World, DebugFrom, DebugTo, FLinearColor((float)r, float(g), float(b), float(alpha)), 120.f, 1.f);
    };
```

> if it works by doing this, you're probably confusing center of mass transform with the regular transform somewhere.

I just checked I'm never calling GetCenterOfMassTransform or GetWorldTransform for the character or bodies except for that `CharShape->Draw` call. 

Movement (and gravity) is a 1:1 copy of what exists in the Samples CharacterVirtual test. So I'm not doing anything weird with setting a position, it's all calculated using `SetLinearVelocity`.

Here's what it looks like if I modify the standing shape to use OffsetShape, this is what I would expect it to look like, (just like the samples example) but without having to call OffsetCenterOfMassShape:
![image](https://github.com/jrouwe/JoltPhysics/assets/1073742/ffd62413-fbfb-4b10-9250-2df42dcd7473)


Disregard the UE5 manniquin being offset from the capsule because it's just reading the Character->GetPosition() data.


Is this just me calling Draw or DrawBodies incorrectly? My character has no problem moving around, which I assume would not be the case if Jolt thought the capsule was half inside the box?

Thanks again for taking the time to respond!

**Reply by jrouwe** [reply, marked as answer]
The way you draw the world is correct. Now that you're mentioning that this is in Unreal it makes a lot more sense. Unreal uses cm units while Jolt uses meter units, so you probably tried to correct for this using:

```CharShape->Draw(..., JPH::Vec3::sReplicate(100.0f), ...);```

However, scaling a shape in Jolt always works in local space of the shape (from the center of mass in this case), so what you're (most likely) doing is passing a GetCenterOfMassTransform() that contains an offset in meters and scaling the shape from its center of mass by a factor 100. That means that the translation in the center of mass is a factor 100 too small and it appears that the capsule is with its center on the ground (in actual fact it is probably 1.35/2 + 0.3 = 0.975 cm above the ground instead of the desired 0.975 meter).

**Reply by wirepair** [reply]
Oh my word this explains so many things! I was wondering why my landscape looked enormous when loaded into the jolt sample! I'll follow the docs guide of [scaling before passing into jolt](https://jrouwe.github.io/JoltPhysics/index.html#conventions-and-limits). Thanks again!
